ruby/red-arrow/ext/arrow/record_batch.cc

```diff
@@ -0,0 +1,610 @@
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::StringArray& array, const int64_t i) {
+    int32_t length;
+    const uint8_t* ptr = array.GetValue(i, &length);
+    // TODO: encoding support
+    return rb::protect([&]{
+      return rb_utf8_str_new(reinterpret_cast<const char*>(ptr), length);
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::Date32Array& array, const int64_t i) {
+    const static int32_t JD_UNIX_EPOCH = 2440588; // UNIX epoch in Julian date
+    return rb::protect([&]{
+      auto raw_value = array.Value(i);
+      auto days_in_julian = raw_value + JD_UNIX_EPOCH;
+      return rb_funcall(rb_cDate, id_jd, 1, LONG2NUM(days_in_julian));
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::Date64Array& array, const int64_t i) {
+    return rb::protect([&]{
+      auto raw_value = array.Value(i);
+      VALUE msec = LL2NUM(raw_value);
+      VALUE sec = rb_rational_new(msec, INT2NUM(1000));
+      VALUE time_value = rb_time_num_new(sec, Qnil);
+      return rb_funcall(time_value, id_to_datetime, 0, 0);
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::Time32Array& array, const int64_t i) {
+    // TODO: must test this function
+    // TODO: unit treatment
+    return rb::protect([&]{
+      auto raw_value = array.Value(i);
+      return LONG2NUM(raw_value);
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::Time64Array& array, const int64_t i) {
+    // TODO: must test this function
+    // TODO: unit treatment
+    return rb::protect([&]{
+      auto raw_value = array.Value(i);
+      return LL2NUM(raw_value);
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::TimestampArray& array, const int64_t i, VALUE scale) {
+    return rb::protect([&]{
+      auto raw_value = array.Value(i);
+      VALUE sec = rb_rational_new(LL2NUM(raw_value), scale);
+      return rb_time_num_new(sec, Qnil);
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::TimestampArray& array, const int64_t i) {
+    const auto& type = arrow::internal::checked_cast<const arrow::TimestampType&>(*array.type());
+    VALUE scale = time_unit_to_scale(type.unit());
+    if (NIL_P(scale)) {
+      rb::protect([]() -> VALUE {
+                    rb_raise(rb_eArgError, "Invalid TimeUnit");
+                    return Qnil;
+                  });
+    }
+    return ConvertValue(array, i, scale);
+  }
+
+  inline VALUE ConvertValue(const arrow::ListArray& array, const int64_t i) {
+    return rb::protect([&]{
+      auto list = array.values()->Slice(array.value_offset(i), array.value_length(i));
+      auto gobj = garrow_array_new_raw(&list);
+      return GOBJ2RVAL(gobj);
+    });
+  }
+
+  inline VALUE ConvertValue(const arrow::StructArray& array, const int64_t i) {
+    const auto* struct_type = array.struct_type();
+    const auto nf = struct_type->num_children();
+    return rb::protect([&] {
+      VALUE record = rb_hash_new();
+      for (int k = 0; k < nf; ++k) {
```

Comment:
`j` will be better than `k` because `j` isn't used yet in this scope.

Reply:
Fixed.